Add "debug" tab in devmode to draw ImGui UI

pierricgimmig opened the pull request:
This is a fairly big change, the main idea is to have a dedicated tab in devmode for
rendering various debug information with ImGui. It can also be used to expose 
interactive parameters, colors, toggles, etc. very easily, often with a single line. 

![image](https://user-images.githubusercontent.com/3687222/97822823-ea205d00-1c6b-11eb-96ad-9f0964974bb3.png)

Reviving multiple OpenGL windows support will be used for UI introspection for example, 
but also for having multiple viewports for the same capture. Some ImGui cleanup has 
been done, but more is needed, mainly in ImGuiOrbit.h/.cpp.

- Only initialize ImGui in devmode
- Remove useless GlPanel.h/.cpp
- Add Render callbacks to GlCanvas 
- ImGui calls need to happen inside a RenderImGui scope

I suggest looking at the change in bulk and *not* by individual commits.

NOTE: The help text doesn't have a background anymore. It was done using ImGui before, but we don't
create an ImGui context for the capture view anymore. I am working on adding a fix.

karupayun commented:
I like it! Seems clearly better to have the debug tab instead of drawing it on Capture Window.

BTW, as it is a big change, I will try yo review it by today but maybe I will do it tomorrow. :)

florian-kuebler commented on OrbitGl/ImGuiOrbit.cpp:
```diff
@@ -728,14 +738,51 @@ void OutputWindow::Draw(const char* title, bool* p_opened, ImVec2* a_Size) {
-  static int i1 = 0;
-  if (ImGui::SliderInt("slider int", &i1, 10, 100)) {
-    GCurrentTimeGraph->SetFontSize(i1);
+  // remove "m_"
```
This feels a bit surprising. At least I would expect a comment what is going on here.

pierricgimmig replied:
yeah I removed the function altogether...

florian-kuebler replied:
Okay, thank!

florian-kuebler commented on OrbitGl/ImGuiOrbit.cpp:
```diff
@@ -728,14 +738,51 @@ void OutputWindow::Draw(const char* title, bool* p_opened, ImVec2* a_Size) {
+}
+
+namespace ImGuiOrbit {
+std::string PrettyVariableName(const char* name) {
```
Are we sure we want such a method? If I could follow correctly so far, this is pretty printing variables, that we expose explicitly to the debug view.
This complete method looks very fragile. I'd rather suggest to either, just expose the original variable names (would be easier to find in an IDE anyways), or when exposing them, also add a description string.

pierricgimmig replied:
yeah I removed the function altogether...

florian-kuebler replied:
Thanks!

florian-kuebler commented on OrbitGl/TextRenderer.cpp:
```diff
@@ -166,8 +166,16 @@ void TextRenderer::AddTextInternal(texture_font_t* font, const char* text, const
   float strWidth = 0.f;
   int minX = INT_MAX;
   int maxX = -INT_MAX;
+  constexpr GLuint indices[6] = {0, 1, 2, 0, 2, 3};
```
Nit: I am not sure, if we have any convention yet. Our clang-tidy settings, complains about c-style arrays and suggest to use std::array instead. I don't have strong opinion on this, but just wanted to mention this.

pierricgimmig replied:
sure, changed it to a std::array

florian-kuebler replied:
Okay, thanks!

florian-kuebler commented on OrbitGl/TimeGraphLayout.cpp:
```diff
@@ -36,47 +36,41 @@ TimeGraphLayout::TimeGraphLayout() {
-  FLOAT_SLIDER_MIN_MAX(scale_, 0.05f, 20.f);
-  ImGui::Checkbox("DrawTrackBackground", &m_DrawTrackBackground);
-  ImGui::End();
+  IMGUI_FLOAT_SLIDER(m_TrackLabelOffsetX);
```
Can you please use this opportunity to update the names here to our style guide?

pierricgimmig replied:
Done.

florian-kuebler replied:
Thanks!

florian-kuebler reviewed (commented):
Thanks Pierric for this change. I left some comments. Mostly nits, and nothing that I would insist in addressing before merging. Let me know what you think of those.

karupayun commented:
I got this crash when I clicked in Misc... Could it be?

Orbit: /home/conan/.conan/data/imgui/1.69/bincrafters/stable/build/77263a2e24ef1df522d09982ea4ecda1262bf522/source_subfolder/imgui.cpp:4495: void CheckStacksSize(ImGuiWindow*, bool): Assertion `*p_backup == current && "PushID/PopID or TreeNode/TreePop Mismatch!"' failed.

pierricgimmig commented on OrbitGl/App.cpp:
```diff
@@ -473,6 +492,15 @@ void OrbitApp::RegisterCaptureWindow(CaptureWindow* capture) {
 }
 
+void OrbitApp::RegisterDebugCanvas(GlCanvas* debug_canvas) {
+  CHECK(debug_canvas_ == nullptr);
```
done above.

florian-kuebler replied:
thanks!

pierricgimmig commented on OrbitGl/App.cpp:
```diff
@@ -473,6 +492,15 @@ void OrbitApp::RegisterCaptureWindow(CaptureWindow* capture) {
 
+void OrbitApp::RegisterDebugCanvas(GlCanvas* debug_canvas) {
+  CHECK(debug_canvas_ == nullptr);
+  constexpr uint32_t kDefaultFontSize = 14;
```
Yes, I added a GetInitialFontSize method to GlCanvas, cleaner now, thanks.

florian-kuebler replied:
thanks

pierricgimmig commented on OrbitGl/CaptureWindow.cpp:
```diff
@@ -639,57 +626,47 @@ void CaptureWindow::RenderUI() {
-    if (time_graph_.GetLayout().DrawProperties()) {
-      NeedsUpdate();
+void CaptureWindow::RenderImGui() {
+  if (ImGui::BeginTabBar("DebugTabBar", ImGuiTabBarFlags_None)) {
```
Normally yes, but this helps understand the hierarchy of ui elements, I prefer to keep it.

florian-kuebler replied:
Okay, fine for me.

pierricgimmig commented on OrbitGl/CaptureWindow.cpp:
```diff
@@ -705,32 +682,23 @@ void ColorToFloat(Color color, float* output) {
+      "Toggle Help: 'H'\n";
+
+  text_renderer_.AddText(help_message, world_x, world_y, GlCanvas::kZValueTextUi,
+                         Color(255, 255, 255, 255), font_size_ * 2);
```
made it clearer by using a constant.

pierricgimmig commented on OrbitGl/GlCanvas.cpp:
```diff
@@ -84,14 +84,24 @@ GlCanvas::GlCanvas(uint32_t font_size)
+    case CanvasType::kDebug:
+      return std::make_unique<GlCanvas>(font_size);
+    default:
+      ERROR("Unhandled canvas type (%d)", canvas_type);
```
Used UNREACHABLE() instead, which aborts.

florian-kuebler replied:
Okay, thanks!

pierricgimmig commented on OrbitGl/CaptureWindow.cpp:
```diff
@@ -705,32 +682,23 @@ void ColorToFloat(Color color, float* output) {
+      "Measure: \"Right Click + Drag\"\n"
+      "Toggle Help: 'H'\n";
+
+  text_renderer_.AddText(help_message, world_x, world_y, GlCanvas::kZValueTextUi,
```
Will do shortly.

pierricgimmig replied:
See below.

florian-kuebler replied:
Thanks

pierricgimmig commented:
> I got this crash when I clicked in Misc... Could it be?

Interesting, I can't repro it, neither on Linux nor Windows. Were on on a specific commit or tip of the branch?

pierricgimmig commented:
@karupayun , @florian-kuebler , I added rounded corner box support:

before (ImGui):

![image](https://user-images.githubusercontent.com/3687222/97950646-72762f00-1d4c-11eb-9f37-2b2b578cf3d3.png)


after (Custom):

![image](https://user-images.githubusercontent.com/3687222/97950621-5c686e80-1d4c-11eb-8c3e-a448b8df8517.png)

I also modified the AddText method so that it returns the exact bounding box of the characters that have been drawn. This can be used to add a perfectly fitting backround box. I the case of the help box, we add a margin to that bounding box, and round the corners.

florian-kuebler reviewed (approved):
LGTM. Thanks!

karupayun commented on OrbitGl/Batcher.h:
```diff
@@ -105,6 +105,13 @@ class Batcher {
   void AddShadedBox(Vec2 pos, Vec2 size, float z, const Color& color,
                     std::shared_ptr<Pickable> pickable,
                     ShadingDirection shading_direction = ShadingDirection::kLeftToRight);
+  void AddRoundedBox(Vec2 pos, Vec2 size, float z, float radius, const Color& color,
```
@reichlfl didn't like to add methods to Batcher who includes several primitives. Only a comment.

karupayun commented on OrbitGl/CaptureWindow.cpp:
```diff
@@ -639,57 +618,50 @@ void CaptureWindow::RenderUI() {
-      // window properly on first draw
-      NeedsRedraw();
-      first_help_draw_ = false;
+    if (ImGui::BeginTabItem("Misc")) {
```
About this "Misc" tab:

- Why we have the button and present the view as a separate window?
- At least in my case, the pop up window starts with a bigger size than it has to be (a lot of empty space), do you know why could it be?

karupayun commented on OrbitGl/TextRenderer.h:
```diff
@@ -25,7 +25,8 @@ class TextRenderer {
   void AddText(const char* a_Text, float a_X, float a_Y, float a_Z, const Color& a_Color,
-               uint32_t font_size, float a_MaxSize = -1.f, bool a_RightJustified = false);
+               uint32_t font_size, float a_MaxSize = -1.f, bool a_RightJustified = false,
+               Vec2* out_text_pos = nullptr, Vec2* out_text_size = nullptr);
```
Nit: I Prefer to have first the out_text arguments and after the RightJustified

karupayun reviewed (approved):
Nice PR Pierric! :)

I like to don't have the debug variables on top of the capture window.
You will have a few issues rebasing this change, sorry for taking too much time.
